File: velocity/src/main/java/ru/privatenull/pnauth/velocity/dialog/VelocityDialogService.java
package ru.privatenull.pnauth.velocity.dialog;

import com.velocitypowered.api.proxy.Player;
import net.kyori.adventure.text.Component;

import java.util.List;
import java.util.Map;

/** Transport boundary for Minecraft dialogs on Velocity. */
public interface VelocityDialogService extends AutoCloseable {
    boolean available();
    void show(Player player, DialogForm form);
    void clear(Player player);

    @Override
    default void close() {
    }

    record DialogForm(Component title, Component notice, List<TextField> fields,
                      Component submitLabel, String actionId) {
        public DialogForm {
            fields = List.copyOf(fields);
        }
    }

    record TextField(String key, Component label, int maxLength) {
    }

    @FunctionalInterface
    interface SubmissionHandler {
        void submit(Player player, String actionId, Map<String, String> values);
    }
}

File: velocity/src/main/kotlin/ru/privatenull/pnauth/velocity/dialog/VelocityDialogService.kt
package ru.privatenull.pnauth.velocity.dialog

import com.velocitypowered.api.proxy.Player
import net.kyori.adventure.text.Component

/** Transport boundary for Minecraft dialogs on Velocity. */
interface VelocityDialogService : AutoCloseable {
    fun available(): Boolean
    fun show(player: Player, form: DialogForm)
    fun clear(player: Player)

    override fun close() {}

    data class DialogForm(
        val title: Component,
        val notice: Component?,
        val fields: List<TextField>,
        val submitLabel: Component,
        val actionId: String
    )

    data class TextField(
        val key: String,
        val label: Component,
        val maxLength: Int
    )

    fun interface SubmissionHandler {
        fun submit(player: Player, actionId: String, values: Map<String, String>)
    }
}
